software/controller/lib/hal/gpio.h

```diff
@@ -34,6 +37,24 @@ Reference abbreviations [RM], [DS], etc are defined in hal/README.md.
 
 namespace GPIO {
 
+// General Purpose I/O
+// [RM] 8.4 GPIO Registers (pg 267)
+struct RegisterStructure {
+  uint32_t mode;                   // Mode register [RM] 8.4.1
+  uint32_t output_type;            // Output type register [RM] 8.4.2
+  uint32_t output_speed;           // Output speed register [RM] 8.4.3
+  uint32_t pullup_pulldown;        // Pull-up/pull-down register [RM] 8.4.4
+  uint32_t input_data;             // Input data register [RM] 8.4.5
+  uint32_t output_data;            // Output data register [RM] 8.4.6
+  uint16_t set;                    // Bit set register [RM] 8.4.7
+  uint16_t clear;                  // Bit reset register [RM] 8.4.7
+  uint32_t flash_lock;             // Configuration lock register [RM] 8.4.8
+  uint32_t alternate_function[2];  // Alternate function low/high register [RM] 8.4.{9,10}
+  uint32_t reset;                  // Reset register [RM] 8.4.11
+};
+
+typedef volatile RegisterStructure Register;
+
 /// \TODO: Move these mappings elsewhere
 enum class Port : uint32_t {
```

Comment:
Same as with registers, I think what we ought to have at some `GPIOBase` scope is just opaque listing of ports but without memory mappings. And then actual memory values are only in the CPP, which could be anything depending on specific processor. Same for `PinMode`, etc.. Yes, I know, more tedious work, but will be less painful portability later.